src/vivarium_census_prl_synth_pop/components/population.py

```diff
@@ -956,3 +943,98 @@ def assign_last_name_ids(self, pop: pd.DataFrame) -> pd.Series:
+        new_reference_persons_relationship_to_old_reference_person = population.loc[
+            new_reference_persons, ["household_id", "relation_to_household_head"]
+        ]
+        population.loc[
+            new_reference_persons, "relation_to_household_head"
+        ] = "Reference person"
+
+        # Update simulants born in simulation as biological children to their mother if mother is new reference person
+        biological_children_idx = population.index[
+            (population["date_of_birth"] > self.start_time)
+            & (
+                population["household_id"].isin(
+                    new_reference_persons_relationship_to_old_reference_person["household_id"]
+                )
+            )
+            & (
+                population["guardian_1"]
+                == population["household_id"].map(
+                    new_reference_persons_relationship_to_old_reference_person.reset_index().set_index(
+                        "household_id"
+                    )[
+                        "index"
+                    ]
+                )
+            )
+        ]
+        population.loc[
+            biological_children_idx, "relation_to_household_head"
+        ] = "Biological child"
+
+        # Update other household members relationship to new reference person
+        for relationship in new_reference_persons_relationship_to_old_reference_person[
+            "relation_to_household_head"
+        ].unique():
+            relationship_map = self.reference_person_update_relationships_map.loc[
+                self.reference_person_update_relationships_map[
+                    "new_reference_person_relationship_to_old_reference_person"
+                ]
+                == relationship
+            ].set_index("relationship_to_old_reference_person")
+
+            household_ids_to_update = (
+                new_reference_persons_relationship_to_old_reference_person.loc[
+                    new_reference_persons_relationship_to_old_reference_person[
+                        "relation_to_household_head"
+                    ]
+                    == relationship,
+                    "household_id",
+                ]
+            )
+            simulants_to_update_idx = (
+                population.index[population["household_id"].isin(household_ids_to_update)]
+                .difference(new_reference_persons_relationship_to_old_reference_person.index)
+                .difference(biological_children_idx)
+            )
+            # Update relationships
+            population.loc[
+                simulants_to_update_idx, "relation_to_household_head"
+            ] = population["relation_to_household_head"].map(
+                relationship_map["relationship_to_new_reference_person"]
+            )
+
+        # Handle extreme edge cases where there would not be a value to map to.
+        population.loc[
+            population["relation_to_household_head"].isnull(), "relation_to_household_head"
```

Comment:
This seems slightly worrying, from the perspective that it could cover up errors if any other part of the simulation was introducing nulls here. Can we limit it to only the simulants who were updated?

Reply:
That seems like a great idea.